### DISCUSSION TITLE
Impulse Spectrum sharing 📬

### DISCUSSION THREAD

**Discussion by ssesselmann** [ssesselmann/impulse#22] [Announcements]
A new version of Impulse has just been pushed to Github, the new version is featuring a public spectrum database where you can publish your spectra for other Impulse users to see with a single click button. A new personal settings panel with easy to understand privacy settings has also been added. 

As a community I believe it is important to start collecting quality data in a common repository. 

NOTE: There have been changes to the settings database, so please rename your **~/impulse_data** folder (if it exists) before launching the program.

Please report any bugs directly to me..

Steven

<img width="750" alt="settings" src="https://github.com/ssesselmann/impulse/assets/75355962/6802670c-db75-4063-a5fb-42b86818debc">

<img width="750" alt="spectra" src="https://github.com/ssesselmann/impulse/assets/75355962/ee635620-e4b9-4fb7-b20d-9a48740e8977">

**Comment by NuclearPhoenixx** [comment]
Hi Steven,

is this kind of API open to other programs or just impulse? Would be neat to integrate this into all kinds of software to have like a central hub for example spectra. It's certainly a neat idea to have.

I'd also like to mention this on my GammaDB (https://gammadb.nuclearphoenix.xyz/), is there a website where users can view all submitted spectra (without having to have impulse) or something like that?

Matthias

**Reply by ssesselmann** [reply]
Matthias, 

Thanks, it was actually your NPES format that gave me the idea to have a shared repository. This is still an early beta version. I need to add more 'inputs' so users can utilise notes and other fields in the NPES schema. 

It is certainly possible to extend this outside Impulse, you can actually see a temporary test page at;

[gammaspectacular/spectra](https://www.gammaspectacular.com/spectra)

Technically possible to set up a registration page for other software to get an API, providing the spectra are uploaded in a readable format. 

Let's make it useful for as many people as possible 👍

Steven

**Reply by NuclearPhoenixx** [reply]
Yeah the NPES format is exactly perfect for stuff like that. JSON is **the** data format for web after all, it's making data handling really easy.

I'm excited to see what you can do with this. It would be really cool if you could funnel all kinds of spectra from all kinds of different programs into a page like yours and then look at them all in one place. And also importing (parts of) spectra back into the MCA software is the obvious next step like you've shown 👍🏻

**Reply by ssesselmann** [reply]
FYI. for requesting the small preview spectra I made a simple function on the server that shrinks any size NPES json file to 255 channels in 'spectrum' [...]. The 'numberOfChannels' field stays unchanged, because it is used on the client side to factor the x_factor = ('numberOfChannels / len('spectrum')),  with x * x_factor the polynomial function should work. Have not pushed that update yet...

**Reply by ssesselmann** [reply]
Just a thought, but if we can collect enough calibrated NPES formatted data, we may eventually be able to train an AI model to recognise spectra, what do you think?

**Reply by NuclearPhoenixx** [reply]
> FYI. for requesting the small preview spectra I made a simple function on the server that shrinks any size NPES json file to 255 channels in 'spectrum' [...]. The 'numberOfChannels' field stays unchanged, because it is used on the client side to factor the x_factor = ('numberOfChannels / len('spectrum')), with x * x_factor the polynomial function should work. Have not pushed that update yet...

Yeah that makes sense if your preview spectra are a lot smaller than the full-fledged ones.

> Just a thought, but if we can collect enough calibrated NPES formatted data, we may eventually be able to train an AI model to recognise spectra, what do you think?

I bet you could do something like this. Sounds like the perfect type of data for some kind of pattern recognition algorithm to me. Great idea actually! 👍🏻

**Comment by NuclearPhoenixx** [comment]
> I bet you could do something like this. Sounds like the perfect type of data for some kind of pattern recognition algorithm to me. Great idea actually! 👍🏻

On second thought, I think it won't actually work as smoothly as expected. Problem is the arguably very bad energy resolution of scintillators, so you get very broad peaks. That is inherently hard to interpret because you are missing a lot of the useful information. I'm sure such an algorithm could easily identify simple spectra like calibration test sources or something with only a few peaks and in the best case spread out over the whole energy region so as to not overlap. But as soon as you get overlap it's not as easy.

Also, it's possible that it would only work on a per system basis on the system it was calibrated on. Because obviously spectra don't all look alike so depending on a user's physical setup, scintillator, calibration and settings, you'd get slightly different results.

I think this would work really well for HPGe detectors where you have practically no overlap of any peaks in almost all circumstances.

Not to say, it's not worth a try. It surely still sounds pretty cool to be honest.
